File: golang/sliding_window/length_sub_string.go
package main

func lengthOfLongestSubString(s string) int {
	charIndex := make(map[byte]int)

	left, maxLength := 0, 0

	for right := 0; right < len(s); right++ {
		if idx, found := charIndex[s[right]]; found && idx >= left {
			left = idx + 1
		}

		charIndex[s[right]] = right

		maxLength = max(maxLength, right-left+1)
	}

	return maxLength
}

File: java/src/main/java/com/sanedge/interview/sliding_window/LengthSubString.java
package com.sanedge.interview.sliding_window;

import java.util.HashMap;
import java.util.Map;

public class LengthSubString {
  public int lengthOfLongestSubString(String s) {
    Map<Character, Integer> charIndex = new HashMap<>();

    int left = 0;
    int maxLength = 0;

    for (int right = 0; right < s.length(); right++) {
      char currentChar = s.charAt(right);

      if (charIndex.containsKey(currentChar) && charIndex.get(currentChar) >= left) {
        left = charIndex.get(currentChar) + 1;
      }

      charIndex.put(currentChar, right);

      maxLength = Math.max(maxLength, right - left + 1);

    }

    return maxLength;

  }

}
